### PR TITLE
vendor: bump runc to rc95

### LINKED ISSUE #88197: kubelet incorrectly configures `kubepods.slice` systemd unit when kube-reserved and/or system-reserved is used
**What happened**:

Kubelet is started on CentOS 7 with systemd with the following switches (among others) `--cgroup-driver=systemd`, `--enforce-node-allocatable=pods`, `--kube-reserved=cpu=100m,ephemeral-storage=2Gi,memory=1.5Gi`

With these settings kubelet correctly creates `kubepods.slice` cgroup with memory limit set to total RAM on the node (~4Gi in this example) minus kube-reserved (1.5Gi in this example)
```
# free -b
              total        used        free      shared  buff/cache   available
Mem:     3973447680   167432192  3489378304      528384   316637184  3584466944

# cat /sys/fs/cgroup/memory/kubepods.slice/memory.limit_in_bytes
2362834944
```

At the same time corresponding systemd unit `kubepods.slice` created by kubelet has a `MemoryLimit` property incorrectly set to total RAM of the node (without subtracting kube-reserved amount):
```
# systemctl show kubepods.slice
Slice=-.slice
ControlGroup=/kubepods.slice
...
MemoryLimit=3973447680
...

# systemctl cat kubepods.slice
...
# /run/systemd/system/kubepods.slice.d/50-MemoryLimit.conf
[Slice]
MemoryLimit=3973447680
...
```

This leads to systemd sometimes resetting the limit in `/sys/fs/cgroup/memory/kubepods.slice/memory.limit_in_bytes` cgroup file to total RAM on the node without accounting for the kube-reserved (and/or system-reserved) amount.
This in turn can lead to pods using more memory than expected and OOM kill events, and unstable node behavior in general

**What you expected to happen**:

`kubepods.slice` systemd unit `MemoryLimit` property should be set to the same value as `/sys/fs/cgroup/memory/kubepods.slice/memory.limit_in_bytes` taking into account kube-reserved and system-reserved amounts.

When kubelet s restarted with different kube-reserved and/or system-reserved values, both `/sys/fs/cgroup/memory/kubepods.slice/memory.limit_in_bytes` and `MemoryLimit` property of `kubepods.slice` systemd unit should be changed to the same value even if the cgroups and the systemd unit already exist.

**How to reproduce it (as minimally and precisely as possible)**:

Run kubelet on any systemd linux with the switches specified above and observe the discrepancy between memory limits defined in `/sys/fs/cgroup/memory/kubepods.slice/memory.limit_in_bytes` and `MemoryLimit` property of `kubepods.slice` systemd unit.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
  1.16.0, but also applies to all earlier and later versions
- Cloud provider or hardware configuration:
   virtualbox VM, but applies to any infrastructure
- OS (e.g: `cat /etc/os-release`):
   centos 7, but applies to any systemd based linux flavor
- Kernel (e.g. `uname -a`):
   `3.10.0-1062.9.1.el7.x86_64 #1 SMP Fri Dec 6 15:49:49 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux` but applies to any other kernel
- Install tools:
   kublr, but applies to any other tool
- Network plugin and version (if this is a network-related bug): N/A
- Others: N/A

### LINKED ISSUE #100328: Kubernetes Pods create/Status unsure (kubelet) - dbus: connection closed by user
#### What happened:
We have a random pod having issues where its status is not available to K8s (Api server) and when we login to assigned node we can see it via ```docker ps ``` in some cases. As of now the solution is we restart kubelet on the node where the pod is assigned 

```
15m         Normal    Scheduled                  pod/catalog-server-5f5cfdffd9-7b8pw    Successfully assigned apps/catalog-server-5f5cfdffd9-7b8pw to node5
4s          Warning   FailedCreatePodContainer   pod/catalog-server-5f5cfdffd9-7b8pw    unable to ensure pod container exists: failed to create container for [kubepods burstable pod3b7a3b5d-7685-43cb-a0ab-5b0646ba07d7] : dbus: connection closed by user
```

```  
Warning  FailedCreatePodContainer  4m40s (x4642 over 16h)  kubelet  unable to ensure pod container exists: failed to create container for [kubepods besteffort poda3e18933-f1c9-4c86-a752-e355e632ef1c] : dbus: connection closed by user
```

#### What you expected to happen: 
This is not happening in our other cluster which have the same containers but older version of kubernetes ranging from 1.14 to 1.17 which were created using kubeadm.

#### Environment:
- Kubernetes version (use `kubectl version`): v1.19.7
- Cloud provider or hardware configuration: 
- OS (e.g: `cat /etc/os-release`): CentOS Linux release 7.9.2009 (Core)
- Kernel (e.g. `uname -a`): 5.4.91-1.el7.elrepo.x86_64
- Install tools: kubespray ( 2.15.0)
- Network plugin and version (if this is a network-related bug): cilium 
- Others:

### PR DESCRIPTION
**Pull request by kolyshkin**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

* runc rc95 contains a fix for CVE-2021-30465, see release notes at
https://github.com/opencontainers/runc/releases/tag/v1.0.0-rc95

* runc rc94 provides some fixes and improvements, see release notes at https://github.com/opencontainers/runc/releases/tag/v1.0.0-rc94

#### Which issue(s) this PR fixes:

https://github.com/kubernetes/kubernetes/issues/101989
Fixes: https://github.com/kubernetes/kubernetes/issues/100328
Fixes: https://github.com/kubernetes/kubernetes/issues/88197

#### Special notes for your reviewer:

This also carries a couple of commits from #100631, and a change from https://github.com/kubernetes/kubernetes/pull/98374.

The previous attempt to bump runc (https://github.com/kubernetes/kubernetes/pull/101888) was reverted (https://github.com/kubernetes/kubernetes/pull/102094) due to a failing test (https://github.com/kubernetes/kubernetes/issues/102092). We believe that picking up https://github.com/kubernetes/kubernetes/pull/98374 fixes that issue; please check
https://prow.k8s.io/?pull=102147&job=*pull-kubernetes-node-crio-e2e

#### Does this PR introduce a user-facing change?
```release-note
Fix a kubelet CPU regression in 1.21 and fixes resource enforcement when using systemd cgroup driver
```

### CODE REVIEW

**Comment by dims** [comment]
@kolyshkin can we please use `v1.0.0-rc95`?

**Comment by kolyshkin** [comment]
Aaand it works!
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102147/pull-kubernetes-node-crio-e2e/1395066014919561216/

Will bump to rc95 now.

**Comment by kolyshkin** [comment]
The test failure that resulted in a revert of a previous bump is now passing:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/102147/pull-kubernetes-node-crio-e2e/1395077376756420608/

**Comment by kolyshkin** [comment]
Fixed the failure in `pull-kubernetes-verify`.

**Comment by liggitt** [comment]
would be good to resolve comment on https://github.com/kubernetes/kubernetes/pull/98374#pullrequestreview-663565908

**Comment by kolyshkin** [comment]
> would be good to resolve comment on [#98374 (review)](https://github.com/kubernetes/kubernetes/pull/98374#pullrequestreview-663565908)

IMHO this can and should be addressed separately, as this is not something that this commit adds (like I originally thought as I overlooked it).

**Comment by kolyshkin** [comment]
/test pull-kubernetes-node-crio-e2e

NOTE we're doing those repeated runs ^^^^ for the sake of having 10-15 clean test runs, to make sure that the issue https://github.com/kubernetes/kubernetes/issues/102092 is not resurfacing. So far all finished runs (but one) are green, and one is failed on a different test case.

The status of these is not reported back to github, but can be seen from https://prow.k8s.io/?pull=102147&job=*pull-kubernetes-node-crio-e2e

**Comment by ehashman** [comment]
/skip
I think the alpha job is broken... no passes: https://prow.k8s.io/job-history/gs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-gce-alpha-features

**Review by ehashman — COMMENTED** [review]
Test output and code changes LGTM but I am not super familiar with the details of the runtimes so leaving others to lgtm/approve

**Comment by odinuge** [comment]
> @ehashman: Test output and code changes LGTM but I am not super familiar with the details of the runtimes so leaving others to lgtm/approve

I think we should preferably get `node-kubelet-serial` going again to verify for any regressions, see https://github.com/kubernetes/kubernetes/issues/102148. We should preferably have a test suite similar running with systemd cgroup driver as well. Maybe @harche could look at that as you have been working with those kinds of things, that would be fantastic!

The interesting part here is cgroup enforcement, for both the cgroupfs and the systemd driver. Currently this is only tested in `node-kubelet-serial`.

**Comment by harche** [comment]
While the `pull-kubernetes-node-crio-e2e` is passing fine, @kolyshkin  I am curious if the[ change to the conformance test ](https://github.com/kubernetes/kubernetes/pull/102147/files#diff-84edb233c83975783af6bff27cc031ff7edb5efb3cf67ff2302f1751ea52500dR127)must for updating runc.

/hold

File: test/e2e_node/summary_test.go [hunk 1]
```diff
@@ -124,7 +124,7 @@ var _ = SIGDescribe("Summary API [NodeConformance]", func() {
 				"WorkingSetBytes": bounded(10*e2evolume.Kb, memoryLimit),
 				"RSSBytes":        bounded(1*e2evolume.Kb, memoryLimit),
 				"PageFaults":      bounded(0, 1000000),
-				"MajorPageFaults": bounded(0, 10),
+				"MajorPageFaults": bounded(0, 1000),
```
Comment (odinuge) [inline test/e2e_node/summary_test.go:127]: Can you remove this commit from the PR, and then we can raise this in another PR/issue? (ref https://github.com/kubernetes/kubernetes/pull/101888#discussion_r634140160)

This is changing a conformance test, and it is for making cgroup v2 tests pass, and I think that is outside the scope of this PR.

**Review by odinuge — COMMENTED** [review]
If the plan is to merge this, I think we at least should keep the changelog entry from https://github.com/kubernetes/kubernetes/pull/98374, since that is a notable thing to mention in the changelog.

**Comment by harche** [comment]
> > @ehashman: Test output and code changes LGTM but I am not super familiar with the details of the runtimes so leaving others to lgtm/approve
> 
> I think we should preferably get `node-kubelet-serial` going again to verify for any regressions, see #102148. We should preferably have a test suite similar running with systemd cgroup driver as well. Maybe @harche could look at that as you have been working with those kinds of things, that would be fantastic!
> 
> The interesting part here is cgroup enforcement, for both the cgroupfs and the systemd driver. Currently this is only tested in `node-kubelet-serial`.

Sounds like a good idea, let me take a closer look at `node-kubelet-serial` test suite.

File: test/e2e_node/summary_test.go (hunk 1 above)
Comment (kolyshkin) [reply test/e2e_node/summary_test.go:127]: Sure; done.

**Comment by kolyshkin** [comment]
> If the plan is to merge this, I think we at least should keep the changelog entry from #98374, since that is a notable thing to mention in the changelog.

@odinuge sure, added.

I am not sure if anything else also deserves a mention; added 2 links to bugs this update I believe fixes.

**Comment by odinuge** [comment]
> @odinuge sure, added.

Thanks!

> I am not sure if anything else also deserves a mention; added 2 links to bugs this update I believe fixes.

You can add https://github.com/kubernetes/kubernetes/issues/88197 as well.

**Comment by kolyshkin** [comment]
> While the pull-kubernetes-node-crio-e2e is passing fine, @kolyshkin I am curious if the change to the conformance test must for updating runc.

@harche commit removed; I don't think it is required -- most probably it just makes a flake less flaky, but @giuseppe may know better.

**Comment by harche** [comment]
> > While the pull-kubernetes-node-crio-e2e is passing fine, @kolyshkin I am curious if the change to the conformance test must for updating runc.
> 
> @harche commit removed; I don't think it is required -- most probably it just makes a flake less flaky, but @giuseppe may know better.

Thanks @kolyshkin 

/hold cancel

**Review by harche — APPROVED** [review]
/lgtm

**Comment by SergeyKanzhelev** [comment]
@kolyshkin can you pls change description so the below bug will not automatically close: 

> Fixes: #101989

So we will track the work of back porting

**Comment by kolyshkin** [comment]
> can you pls change description so the below bug will not automatically close:

Done

**Comment by bobbypage** [comment]
> I think we should preferably get node-kubelet-serial going again to verify for any regressions, see #102148. We should preferably have a test suite similar running with systemd cgroup driver as well. Maybe @harche could look at that as you have been working with those kinds of things, that would be fantastic!

another idea in the meantime is we can run this PR against subset of those tests manually for now. I think the main tests that look relevant are `node_container_manager_test` and `container_manager_test.go`

**Review by BenTheElder — APPROVED** [review]
vendor changes are LGTM

**Comment by ehashman** [comment]
Re: https://github.com/kubernetes/kubernetes/pull/102147#issuecomment-844756376 I opened https://github.com/kubernetes/test-infra/pull/22284 to add that job as a presubmit for cases like this.

Filed #102194 for failing alpha job.
/skip

**Comment by liggitt** [comment]
This has the runc and test changes needed to pick back to 1.21?

**Comment by ehashman** [comment]
> This has the runc and test changes needed to pick back to 1.21?

- Test changes: d312ef7eb6730099607627196888c4f91044bb86 from #98374
- runc bump: f3cdfc488e88714f2fe695e35ad0d3f8e45ddc25

Next 2 commits make the new API work.

**Pull request merged**